src/solvers/dg_multi/types_and_traits.jl

```diff
@@ -0,0 +1,49 @@
+
+# `MultiDG` refers to both multiple DG types (polynomial/SBP, simplices/quads/hexes) as well as
+# the use of multi-dimensional operators in the solver.
+const MultiDG{NDIMS, ElemType, ApproxType, SurfaceIntegral, VolumeIntegral} =
+  DG{<:RefElemData{NDIMS, ElemType, ApproxType}, Mortar, SurfaceIntegral, VolumeIntegral} where {Mortar}
+
+# these are necessary for pretty printing
+polydeg(dg::MultiDG) = dg.basis.N
+Base.summary(io::IO, dg::DG) where {DG <: MultiDG} = print(io, "MultiDG(polydeg=$(polydeg(dg)))")
+Base.real(rd::RefElemData{NDIMS, Elem, ApproxType, Nfaces, RealT}) where {NDIMS, Elem, ApproxType, Nfaces, RealT} = RealT
+
+"""
+    MultiDG(; polydeg::Integer,
+              elem_type::AbstractElemShape,
+              approximation_type=Polynomial(),
+              surface_flux=flux_central,
+              surface_integral=SurfaceIntegralWeakForm(surface_flux),
+              volume_integral=VolumeIntegralWeakForm())
+
+Create a discontinuous Galerkin method which uses
+- approximations of polynomial degree `polydeg`
+- element type `elem_type` (`Tri()`, `Quad()`, `Tet()`, and `Hex()` currently supported)
+
+Optional:
+- approximation type of `approximation_type` (default is `Polynomial()`; `SBP()` also supported for
+  `Tri()`, `Quad()`, and `Hex()` element types).
+"""
+function MultiDG(; polydeg::Integer,
+                   elem_type::AbstractElemShape,
+                   approximation_type=Polynomial(),
+                   surface_flux=flux_central,
+                   surface_integral=SurfaceIntegralWeakForm(surface_flux),
+                   volume_integral=VolumeIntegralWeakForm())
+  rd = RefElemData(elem_type, approximation_type, polydeg)
+  return DG(rd, nothing #= mortar =#, surface_integral, volume_integral)
+end
+
+# type aliases for dispatch purposes
+const MultiDGWeakForm{NDIMS, ElemType, ApproxType} =
+  MultiDG{NDIMS, ElemType, ApproxType, <:SurfaceIntegralWeakForm, <:VolumeIntegralWeakForm}
+
+const PolyDGFluxDiff{NDIMS, ElemType} =
+  MultiDG{NDIMS, ElemType, Polynomial, <:SurfaceIntegralWeakForm, <:VolumeIntegralFluxDifferencing} where {NDIMS, ElemType}
+
+const SBPDGFluxDiff{Dim, Elem} =
+  MultiDG{NDIMS, ElemType, <:SBP, <:SurfaceIntegralWeakForm, <:VolumeIntegralFluxDifferencing} where {NDIMS, ElemType}
```

Comment:
Alternatively, we could also get rid of these additional types and use the same approach as in the rest of Trixi: Whenever we dispatch on a sub-structure of something, pass it as additional argument to make dispatch more obvious. For example, we extract the `volume_integral` from `dg` and pass `volume_integral, dg` to 
```julia
calc_volume_integral!(du, u, mesh, nonconservative_terms, equations, volume_integral, dg, cache)
```
Related to this: Could you please use the same [signature convention](https://trixi-framework.github.io/Trixi.jl/dev/styleguide/) that we use in the rest of Trixi:
- Use the order `mesh, equations, solver, cache`
- If something needs to be specified in more detail for dispatch, put the additional argument before the general one that is specified in more detail. For example, we use `have_nonconservative_terms(equations), equations` and `dg.volume_integral, dg`.

Reply:
> Whenever we dispatch on a sub-structure of something, pass it as additional argument to make dispatch more obvious.

That is a good point. I think it's my fault that @jlchan introduced these types, because originally there were a lot of very long and hard to comprehend type annotations, and I suggested to use simpler types to capture the motivation here. However, for consistency and clarity I like your approach much better - even more so as these additional types are not used for anything but dispatching.

One question though: When we pass additional arguments for dispatch, do we also want to do this where we do not actually need the *value* of the argument inside the function? And if this would be the case, would it bother us to have function arguments that are proper variables (and not just traits like `have_nonconservative_terms`) but that are never used inside the function?

I am thinking about the case where we pass `volume_integral` just to make sure that we pick up the correct method, even though `volume_integral` is not used. Not sure where I stand on this (= no hard opinion).

Reply:
> When we pass additional arguments for dispatch, do we also want to do this where we do not actually need the _value_ of the argument inside the function? And if this would be the case, would it bother us to have function arguments that are proper variables (and not just traits like `have_nonconservative_terms`) but that are never used inside the function?

I you wanted to get rid of this, we would need to throw away all but one mesh type (or restructure everything significantly). For example, the `mesh` is usually not used in something like `calc_volume_integral!` but needed for dispatch. The same is true for singletons like `VolumeIntegralWeakForm`. Thus, I would say it is completely okay to pass arguments used for dispatch.

Reply:
@ranocha I think I may have copied the signature from an unstructured mesh solver file (which may have been updated since). I'll make sure to fix that - thanks for the catch!

@sloede ok - I'll prune the extra types in the next commit. Thanks for the explanation!

Reply:
No worries - we didn't have a unified signature of all this stuff not long ago but have been trying to improve this

Reply:
> Thus, I would say it is completely okay to pass arguments used for dispatch.

Thanks for the clarification. I agree, if we're doing this already (especially for the mesh), then there's no reason why we shouldn't do it for other arguments as well.

Reply:
So just to clarify - if we dispatch on a type parameter, we will generally unroll that into an extra argument to make dispatch more obvious?

Reply:
I tried explicitly "unpacking" the type parameters for dispatch, but it seems less clean to me than the type alias approach. I think this is mostly due to the fact that I dispatch on `Union{DGMultiWeakForm, DGMultiFluxDiff{<:SBP}}`, which seems cleaner using type aliases. 

I replaced the non-intuitive types with `DGMultiWeakForm{...}` and `DGMultiFluxDiff{...}`, so that `PolyDGFluxDiff = DGMultiFluxDiff{Polynomial}` and `SBPDGFluxDiff = DGMultiFluxDiff{SBP}`. @sloede @ranocha would this be OK for this PR?

Reply:
Yes, sounds also okay to me :+1: